File: PassByValue.java
// to run:  javac PassByValue.java && java PassByValue
class PassByValue{
    public static void main(String[] args){
        emp a=new emp("a");
        emp b=new emp("b");
        foo(a,b);
        System.out.println(a);
        System.out.println(b);

        char ac='a';
        char bc='b';
        bar(ac,bc);
        System.out.println(ac);
        System.out.println(bc);

    }

    private static void foo(emp a,emp b){
        a.setName("1");
        b=new emp("2");
        System.out.println("This is inside foo method "+b);
    }

    private static void bar(char a,char b){
        a = 'q';
        b = 'e';
    }
}

class emp{
    String name;
    public emp(String name){
        this.name=name;
    }
    public void setName(String name){
        this.name=name;
    }
    public String toString(){
        return name;
    }
}

File: PassByValue.cpp
#include <iostream>
using namespace std;
//g++ PassByValue.cpp && ./a.out

class emp{
    public:
    char name;
    emp(char nameParam){
        name=nameParam;
    }
    void setName(char nameParam){
        name=nameParam;
    }
    char toString(){
        return name;
    }
};

void foo(emp a,emp b){
    a.setName('1');
    b= emp('2');
}

void fooRef(emp &a,emp &b){
    a.setName('1');
    b= emp('2');
}

int main()
{
    emp a('a');
    emp b('b');
    foo(a,b);
    cout << "after passbyVal a = " << a.toString() << " b = " << b.toString() << "\n";
    fooRef(a,b);
    cout << "after passByRef a = " << a.toString() << " b = " << b.toString() << "\n";
}